xonsh/tools.py

```diff
@@ -92,6 +93,85 @@ def __init__(self, returncode, command, output=None, stderr=None,
         self.completed_command = completed_command
 
 
+class EnvPath(MutableSequence):
+    """
+    A class that implements an environment path, which is a list of
+    strings. Provides a custom method that expands all paths if the
+    relevant env variable has been set.
+    """
+    @staticmethod
+    def _expandpath(path):
+        """
+        Performs environment variable / user expansion on a given path
+        if the relevant flag has been set.
+        """
+        env = getattr(builtins, '__xonsh_env__', os.environ)
+        if env.get('EXPAND_ENV_VARS', False):
+            # expand variables and use os.path.realpath to handle cases
+            # with relative paths like ../ or ./
+            path = os.path.realpath(os.path.expanduser(expandvars(path)))
+        return path
+
+    @staticmethod
+    def _decode_bytes(path):
+        """
+        Tries to decode a path in bytes using XONSH_ENCODING if available,
+        otherwise using sys.getdefaultencoding().
+        """
+        env = getattr(builtins, '__xonsh_env__', os.environ)
+        enc = env.get('XONSH_ENCODING', sys.getdefaultencoding())
+        return path.decode(encoding=enc)
+
+    def __init__(self, *args):
+        self._d = []
+        if len(args) > 0:
+            if isinstance(args[0], str):
+                self._d = args[0].split(os.pathsep)
+            elif isinstance(args[0], pathlib.Path):
+                self._d = [args[0]]
+            elif isinstance(args[0], bytes):
+                # decode bytes to a string and then split based on
+                # the default path separator
+                self._d = EnvPath._decode_bytes(args[0]).split(os.pathsep)
+            elif isinstance(args[0], list):
+                self._d = list(args[0])
```

Comment:
@scopatz What's the best type of exception to raise in a constructor? I'm thinking either `ValueError` or `TypeError`.